jetty-server/src/main/java/org/eclipse/jetty/server/DetectorConnectionFactory.java

```diff
@@ -0,0 +1,322 @@
+                        throw new IllegalStateException("Cannot upgrade: " + nextConnection + " does not implement " + Connection.UpgradeTo.class.getName());
+                    }
+                    getEndPoint().upgrade(nextConnection);
+                    _connector.getByteBufferPool().release(_buffer);
+                    return true;
+                }
+                noMagicFound &= detection == Detecting.Detection.NOT_RECOGNIZED;
+            }
+
+            if (noMagicFound)
+            {
+                // no DetectingConnectionFactory recognized those bytes -> call no recognition callback
+                _detectionUnsuccessful.onDetectionUnsuccessful(DetectorConnectionFactory.this, _connector, getEndPoint(), _buffer);
+                return true;
+            }
+            return false;
+        }
+
+        private void releaseAndClose()
+        {
+            _connector.getByteBufferPool().release(_buffer);
+            close();
+        }
+
+        @Override
+        public void onUpgradeTo(ByteBuffer prefilled)
+        {
+            if (BufferUtil.hasContent(prefilled))
+                BufferUtil.append(_buffer, prefilled);
+        }
+
+        @Override
+        public ByteBuffer onUpgradeFrom()
+        {
+            return _buffer;
+        }
+    }
+
+    /**
+     * A functional interface whose instances are used as callbacks when protocol detection is unsuccessful.
+     */
+    @FunctionalInterface
+    interface DetectionUnsuccessful
+    {
+        /**
+         * This method is called back when protocol detection is unsuccessful. Instances of this interface are responsible
+         * for disposing of the {@code buffer} and the {@code endpoint}.
+         * @param connector the connector.
+         * @param endPoint the endpoint.
+         * @param buffer the buffer containing the bytes.
+         */
+        void onDetectionUnsuccessful(Connector connector, EndPoint endPoint, ByteBuffer buffer);
+    }
+
+    /**
+     * Equivalent to {@link DetectionUnsuccessful} but used internally to allow subclasses to pass to {@code super} an instance
+     * that can callback a member method.
+     * <p>Only subclasses of {@link DetectorConnectionFactory} who require calling back a member method need this interface.</p>
+     */
+    @FunctionalInterface
+    protected interface DetectionUnsuccessfulTrampoline
+    {
+        /**
+         * This method is called back when protocol detection is unsuccessful. Instances of this interface are responsible
+         * for disposing of the {@code buffer} and the {@code endpoint}.
+         * @param connectionFactory the detector that performed the detection.
+         * @param connector the connector.
+         * @param endPoint the endpoint.
+         * @param buffer the buffer containing the bytes.
+         */
+        void onDetectionUnsuccessful(DetectorConnectionFactory connectionFactory, Connector connector, EndPoint endPoint, ByteBuffer buffer);
+    }
```

Comment:
So the problem is that we cannot pass a function to the constructor that references `this`, so we need `DetectionSuccessful` for the function, but then `DetectionUnsuccessfulTrampoline` to workaround referencing `this`.

This to me means that it's the wrong solution. If we need to reference `this`, then all of the above should be part of the subclass itself and so an overridable method can do the job.
The overridable method will get rid of these 2 interfaces and all the machinery that they require.

Reply:
Okay, I'll do this.